plugin/healthchecker/filter.go

```diff
@@ -0,0 +1,115 @@
+package healthchecker
+
+import (
+	"time"
+
+	"github.com/bluele/gcache"
+	"github.com/miekg/dns"
+)
+
+type (
+	HealthCheckFilter struct {
+		cache    gcache.Cache
+		checker  Checker
+		interval time.Duration
+		names    map[string]struct{}
+	}
+
+	entry struct {
+		dnsRecord dns.RR
+		healthy   bool
+		quit      chan bool
+	}
+)
+
+func NewHealthCheckFilter(checker Checker, size, interval int, names map[string]struct{}) *HealthCheckFilter {
+	return &HealthCheckFilter{
+		cache: gcache.New(size).LRU().EvictedFunc(func(key interface{}, val interface{}) {
+			if e, ok := val.(entry); ok {
+				e.quit <- true
+			}
+		}).Build(),
+		checker:  checker,
+		interval: time.Duration(interval) * time.Millisecond,
+		names:    names,
+	}
+}
+
+func (p *HealthCheckFilter) FilterRecords(records []dns.RR) []dns.RR {
+	if p.cache.Len(false) == 0 {
+		recs := make([]dns.RR, len(records))
+		copy(recs, records)
+		go func() {
+			for _, r := range recs {
+				if _, ok := p.names[r.Header().Name]; ok || len(p.names) == 0 {
+					p.put(r)
+				}
+			}
+		}()
+		return records
+	}
+
+	result := make([]dns.RR, 0, len(records))
+
+	for _, r := range records {
+		if _, ok := p.names[r.Header().Name]; ok || len(p.names) == 0 {
+			e := p.get(r.String())
+			if e != nil {
+				if e.healthy {
+					result = append(result, e.dnsRecord)
+				}
+				continue
+			}
+			p.put(r)
+		}
+		result = append(result, r)
+	}
+
+	return result
+}
+
+func (p *HealthCheckFilter) put(rec dns.RR) {
+	health := p.checker.Check(rec)
+	quit := make(chan bool)
+	err := p.cache.Set(rec.String(), entry{
+		dnsRecord: rec,
+		healthy:   health,
+		quit:      quit,
+	})
+	if err != nil {
+		log.Warningf("could not put an entry to cache: %s", rec.String())
+	}
+
+	ticker := time.NewTimer(p.interval)
+	go func() {
+		for {
+			select {
+			case <-quit:
+				return
+			case <-ticker.C:
```

Comment:
Can we avoid using `quit` channel? Every tick we can check record presence in the cache and if it doesn't exist just stop ticker.

Reply:
`quit` channel does not provide much overhead for `entry` structure but provides immediate reaction at entry eviction from the cache. Otherwise with long ticker duration, routine may live much longer than it should.